cedar-policy/src/ffi/utils.rs

```diff
@@ -17,12 +17,134 @@
+pub struct DetailedError {
+    /// Main error message, including both the `miette` "message" and the
+    /// `miette` "causes" (uses `miette`'s default `Display` output)
+    pub message: String,
+    /// Help message, providing additional information about the error or help resolving it
+    pub help: Option<String>,
+    /// Error code
+    pub code: Option<String>,
+    /// URL for more information about the error
+    pub url: Option<String>,
+    /// Severity
+    pub severity: Option<Severity>,
+    /// Source labels (ranges)
+    #[serde(default)]
+    pub source_locations: Vec<SourceLabel>,
+    /// Related errors
+    #[serde(default)]
+    pub related: Vec<DetailedError>,
+}
+
+/// Exactly like `miette::Severity` but implements `Hash`
+///
+/// If `miette::Severity` adds `derive(Hash)` in the future, we can remove this
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash, Deserialize, Serialize)]
+pub enum Severity {
+    /// Advice (the lowest severity)
+    Advice,
+    /// Warning
+    Warning,
+    /// Error (the highest severity)
+    Error,
+}
+
+impl From<miette::Severity> for Severity {
+    fn from(severity: miette::Severity) -> Self {
+        match severity {
+            miette::Severity::Advice => Self::Advice,
+            miette::Severity::Warning => Self::Warning,
+            miette::Severity::Error => Self::Error,
+        }
+    }
+}
+
+/// Structure of the JSON output representing a `miette` source label (range)
+#[derive(Debug, PartialEq, Eq, Clone, Hash, Deserialize, Serialize)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(rename_all = "camelCase")]
+pub struct SourceLabel {
+    /// Text of the label (if any)
+    pub label: Option<String>,
+    /// Source location (range) of the label
+    #[serde(flatten)]
+    pub loc: SourceLocation,
+}
+
+/// A range of source code representing the location of an error or warning.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Hash, Deserialize, Serialize)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(rename_all = "camelCase")]
+pub struct SourceLocation {
+    /// Start of the source location (in bytes)
+    pub start: usize,
+    /// End of the source location (in bytes)
+    pub end: usize,
+}
+
+impl From<miette::LabeledSpan> for SourceLabel {
+    fn from(span: miette::LabeledSpan) -> Self {
+        Self {
+            label: span.label().map(ToString::to_string),
+            loc: SourceLocation {
+                start: span.offset(),
+                end: span.offset() + span.len(),
+            },
+        }
+    }
+}
+
+impl<'a, E: miette::Diagnostic + ?Sized> From<&'a E> for DetailedError {
+    fn from(diag: &'a E) -> Self {
+        Self {
+            message: {
+                let mut s = diag.to_string();
+                let mut source = diag.source();
```

Comment:
What is `.source()`? Not related to source locations?

Reply:
It's a method on the [`Error` trait in std](https://doc.rust-lang.org/std/error/trait.Error.html).  I'm not sure why, but I needed to add these `:` and `source()` in order to preserve previous behavior (where this is the string message that was previously passed in the FFI).